net_sink.go

```diff
@@ -305,34 +304,43 @@ func (s *netSink) run() {
 			n := len(s.outc)
 			for i := 0; i < n && s.conn != nil; i++ {
 				buf := <-s.outc
-				s.writeToConn(buf)
+				if err := s.writeToConn(buf); err != nil {
+					s.retryc <- buf
+					continue
+				}
 				putBuffer(buf)
 			}
 			close(done)
 		case buf := <-s.outc:
-			s.writeToConn(buf)
+			if err := s.writeToConn(buf); err != nil {
+				s.retryc <- buf
+				continue
+			}
+			putBuffer(buf)
+		case buf := <-s.retryc:
```

Comment:
you might want to bring this out of this select and into another select right above this one, so that we always check the retry channel first, before we do anything else. Otherwise, I don't think there's a guarantee (select doesn't guarantee order, its random eval) that we pull the stuff out of the retry channel in a reasonable time. I know it's fine to process metric data out of order, though, so maybe this is ok.

My idea is to have this above:
```go
select {
case buf := <-s.retryc:
  // handle buffers that need to be retried first
default:
  // drop through in case retryc has nothing
}
```

Reply:
There may be a case where the message fails for some reason (maybe too big?) so maybe you don't want to always try retryc first.

Reply:
I think it's reasonable to try that first. Giving it a shot.